File: backend/routers/body_metrics.py
from fastapi import APIRouter, Depends, HTTPException, Query, Response, status
from sqlalchemy.orm import Session
from datetime import date

from backend.auth import CurrentUser, get_current_user
from backend.database import get_db
from backend.schemas.body_metric import BodyMetric as BodyMetricSchema, BodyMetricCreate
from backend.services import body_metric_service
from backend.services.exceptions import BodyMetricNotFound, PermissionDenied

router = APIRouter(tags=["Body Metrics"])

@router.get("/users/{userId}/body-metrics", response_model=list[BodyMetricSchema])
def list_body_metrics(
    userId: int,
    from_date: date | None = Query(default=None, alias="from"),
    to_date: date | None = Query(default=None, alias="to"),
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    try:
        return body_metric_service.list_metrics(
            db=db,
            user_id=current_user.id,
            user_role=current_user.role,
            target_user_id=userId,
            from_date=from_date,
            to_date=to_date
        )
    except PermissionDenied as e:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(e))

@router.post("/users/{userId}/body-metrics", response_model=BodyMetricSchema)
def create_body_metric(
    userId: int,
    payload: BodyMetricCreate,
    response: Response,
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    try:
        metric, was_created = body_metric_service.create_metric(
            db=db,
            user_id=current_user.id,
            target_user_id=userId,
            payload=payload
        )
        response.status_code = status.HTTP_201_CREATED if was_created else status.HTTP_200_OK
        return metric
    except PermissionDenied as e:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(e))

@router.get("/users/{userId}/body-metrics/{metricId}", response_model=BodyMetricSchema)
def get_body_metric(
    userId: int,
    metricId: int,
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    try:
        return body_metric_service.get_metric(
            db=db,
            metric_id=metricId,
            user_id=current_user.id,
            user_role=current_user.role,
            target_user_id=userId
        )
    except BodyMetricNotFound as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except PermissionDenied as e:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(e))

@router.put("/users/{userId}/body-metrics/{metricId}", response_model=BodyMetricSchema)
def update_body_metric(
    userId: int,
    metricId: int,
    payload: BodyMetricCreate,
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    try:
        return body_metric_service.update_metric(
            db=db,
            metric_id=metricId,
            user_id=current_user.id,
            target_user_id=userId,
            payload=payload
        )
    except BodyMetricNotFound as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except PermissionDenied as e:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(e))

@router.delete("/users/{userId}/body-metrics/{metricId}", status_code=status.HTTP_204_NO_CONTENT)
def delete_body_metric(
    userId: int,
    metricId: int,
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    try:
        body_metric_service.delete_metric(
            db=db,
            metric_id=metricId,
            user_id=current_user.id,
            user_role=current_user.role,
            target_user_id=userId
        )
    except BodyMetricNotFound as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except PermissionDenied as e:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(e))

File: frontend/src/api/bodyMetrics.ts
import { z } from 'zod'
import { BodyMetricSchema, BodyMetricCreateSchema } from '@/schemas'
import { apiFetch, type Credentials } from './client'

export function listBodyMetrics(
  userId: number,
  credentials: Credentials,
  filters?: { from?: string; to?: string }
) {
  const params = new URLSearchParams()
  if (filters?.from) params.append('from', filters.from)
  if (filters?.to) params.append('to', filters.to)

  const query = params.toString()
  return apiFetch(`/users/${userId}/body-metrics${query ? `?${query}` : ''}`, z.array(BodyMetricSchema), credentials)
}

export function getBodyMetric(userId: number, metricId: number, credentials: Credentials) {
  return apiFetch(`/users/${userId}/body-metrics/${metricId}`, BodyMetricSchema, credentials)
}

export function upsertBodyMetric(userId: number, payload: z.infer<typeof BodyMetricCreateSchema>, credentials: Credentials) {
  // Backend uses POST for upsert (returns 200 if updated, 201 if created)
  return apiFetch(`/users/${userId}/body-metrics`, BodyMetricSchema, credentials, {
    method: 'POST',
    body: JSON.stringify(BodyMetricCreateSchema.parse(payload)),
  })
}

export function updateBodyMetric(userId: number, metricId: number, payload: z.infer<typeof BodyMetricCreateSchema>, credentials: Credentials) {
  return apiFetch(`/users/${userId}/body-metrics/${metricId}`, BodyMetricSchema, credentials, {
    method: 'PUT',
    body: JSON.stringify(BodyMetricCreateSchema.parse(payload)),
  })
}

export function deleteBodyMetric(userId: number, metricId: number, credentials: Credentials) {
  return apiFetch(`/users/${userId}/body-metrics/${metricId}`, z.undefined(), credentials, {
    method: 'DELETE',
  })
}
